### ISSUE TITLE
Error when using Union as constraint for TypeVar

### ISSUE THREAD

**Issue by cdce8p** [microsoft/pylance-release#1412]
<h3> Environment data </h3>

-   Language Server version: 2021.6.0
-   OS and version: darwin x64
-   Python version: 3.10.0b2
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

```python
from typing import TypeVar, Union

class A: ...
class B: ...
class C: ...
class D: ...

T = TypeVar("T", A, B, Union[C, D])


def do_something(value: T) -> T:
    ...

def func(value: Union[C, D]):
    value = do_something(value)  # error
    reveal_type(value)
```

<h3> Expected behaviour </h3>

No error. I'm not completely sure that's the way to do it, but it seems to work with `mypy`
```
test.py:16: note: Revealed type is "Union[test.C, test.D]"
```

<h3> Actual behaviour </h3>

```
Expression of type "T@do_something" cannot be assigned to declared type "C | D"
  Type "T@do_something" cannot be assigned to type "C | D"
    Type "A* | B* | C* | D*" cannot be assigned to type "C"
    Type "A* | B* | C* | D*" cannot be assigned to type "D"
    Type "A* | B* | C* | D*" cannot be assigned to type "C | D"
      "A*" is incompatible with "C"
      "B*" is incompatible with "C"
      "D*" is incompatible with "C"
      "A*" is incompatible with "D"
  ...

Argument of type "C | D" cannot be assigned to parameter "value" of type "T@do_something" in function "do_something"
  Type "C | D" is incompatible with constrained type variable "T"
```

<h3> Logs </h3>
--

**Comment by erictraut** [comment]
Thanks for the bug report. Yes, the type checker logic wasn't properly dealing with the case where a constrained TypeVar included both union and non-union constraints. This is now working, and the fix will be included in the next release.

**Comment by cdce8p** [comment]
Thanks for the quick fix 🚀

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.6.1, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202161-9-june-2021
